opt_adapt/opt.py

```diff
@@ -110,8 +112,8 @@ def _BFGS(forward_run, m, params, u, u_, dJ_, B, Rspace=False):
             else:
                 dJ_ = fd.Function(dJ).assign(dJ_)
                 u_ = fd.Function(u).assign(u_)
-                s = u - u_
-                y = dJ - dJ_
+                s = fd.assemble(u - u_)
```

Comment:
It would be better to use
```
s = u.copy(deepcopy=True)
s -= u_
```
The first part will make `s` a copy of `u` and the second part will subtract `u_`.